### PR TITLE
Register an injection with an initializer.

### PR DESCRIPTION
**Pull request by Heromyth**

### CODE REVIEW

File: source/poodinis/container.d [hunk 1]
```diff
@@ -391,16 +438,16 @@ synchronized class DependencyContainer {
...
-            mixin(createImportsString!Type);
-
-            static if (__traits(compiles, __traits(getProtection, __traits(getMember, instance, memberName)))
-                        && __traits(getProtection, __traits(getMember, instance, memberName)) == "public"
-                        && isFunction!(mixin(fullyQualifiedName!Type ~ `.` ~ memberName))
-                        && hasUDA!(__traits(getMember, instance, memberName), PostConstruct)) {
-                __traits(getMember, instance, memberName)();
-            }
-        }
+        // foreach (memberName; __traits(allMembers, Type)) {
```
Comment (mbierlee) [inline source/poodinis/container.d]: This code is commented out, why?

Comment (Heromyth) [reply source/poodinis/container.d]: It's my mistake. It fixed now. In my tests, the mixin() will lead some strange errors, so it removed.

**Comment by Heromyth** [comment]
Here are some references:
1. https://laravel.com/docs/6.x/providers#the-register-method
2. https://github.com/Swinject/Swinject/blob/master/Documentation/InjectionPatterns.md

File: example/postconstructorpredestructor/app.d [hunk 2]
```diff
@@ -42,8 +42,14 @@ class AClass {
 
 public void main() {
 	auto container = new shared DependencyContainer();
-	container.register!ADependency;
-	container.register!AClass;
+	container.register!(ADependency).onConstructed((Object obj) {
+		writeln("ADependency constructed");
```
Comment (SingingBush) [inline example/postconstructorpredestructor/app.d]: I use poodinis for my project and would rather see use of the standard logger than a writeln. For support across compiler versions it should be wrapped, something like:

```d
    static if(__traits(compiles, (){ import std.experimental.logger; } )) {
        sharedLog.info("ADependency contructed");
    }
```

doing this way allows end users to have control over whether they see the message based on log level and also allows it to be logged to various output

Comment (Heromyth) [reply example/postconstructorpredestructor/app.d]: Yes, we also recommend that the logger should be used to replace the simple writeln().
That seems another PR needed .

Comment (mbierlee) [reply example/postconstructorpredestructor/app.d]: These writelns are currently in examples. Examples should be kept as simple as possible. There's no need to use the logger API there, that's not the goal of the example.

Comment (mbierlee) [reply example/postconstructorpredestructor/app.d]: These writelns are currently in examples. Examples should be kept as simple as possible. There's no need to use the logger API there, that's not the goal of the example.

**Comment by mbierlee** [comment]
Could you add the new features to the tutorial.MD?

**Comment by mbierlee** [comment]
Could you rebase this PR on the "develop" branch?

File: example/postconstructorpredestructor/app.d (hunk 2 above)
Comment (SingingBush) [reply example/postconstructorpredestructor/app.d]: that's fine for this file but there's similar in _source/poodinis/container.d_ I chose bad place to put the comment but meant as a general principal. It's good if logging is handled by something like:

```d
    static if(__traits(compiles, (){ import std.experimental.logger; } )) {
        sharedLog.info("blah blah blah");
    } else if(CHECK FOR VIBE-D) {
        // use vibe log
    } else {
        writeln("blah blah blah");
    }
```

as it allows for your end users to decide how they want log output.

**Comment by Heromyth** [comment]
> 
> 
> Could you rebase this PR on the "develop" branch?

Submitted again on branch develop. See https://github.com/mbierlee/poodinis/pull/28.

**Comment by Heromyth** [comment]
> 
> 
> Could you add the new features to the tutorial.MD?

We will have a try

**Pull request closed without merging**